src/server/pps/server/githook/githook.go

```diff
@@ -0,0 +1,184 @@
+package githook
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/pachyderm/pachyderm/src/client"
+	"github.com/pachyderm/pachyderm/src/client/pps"
+	col "github.com/pachyderm/pachyderm/src/server/pkg/collection"
+	"github.com/pachyderm/pachyderm/src/server/pkg/ppsdb"
+	"github.com/pachyderm/pachyderm/src/server/pkg/util"
+
+	etcd "github.com/coreos/etcd/clientv3"
+	logrus "github.com/sirupsen/logrus"
+	"golang.org/x/net/context"
+	"gopkg.in/go-playground/webhooks.v3"
+	"gopkg.in/go-playground/webhooks.v3/github"
+)
+
+// GitHookPort specifies the port the server will listen on
+const GitHookPort = 999
+const apiVersion = "v1"
+
+// gitHookServer serves GetFile requests over HTTP
+type gitHookServer struct {
+	hook       *github.Webhook
+	client     *client.APIClient
+	etcdClient *etcd.Client
+	pipelines  col.Collection
+}
+
+// RunGitHookServer starts the webhook server
+func RunGitHookServer(address string, etcdAddress string, etcdPrefix string) error {
+	c, err := client.NewFromAddress(address)
+	if err != nil {
+		return err
+	}
+	etcdClient, err := etcd.New(etcd.Config{
+		Endpoints:   []string{etcdAddress},
+		DialOptions: client.EtcdDialOptions(),
+	})
+	if err != nil {
+		return err
+	}
+	hook := github.New(&github.Config{})
+	s := &gitHookServer{
+		hook,
+		c,
+		etcdClient,
+		ppsdb.Pipelines(etcdClient, etcdPrefix),
+	}
+	hook.RegisterEvents(s.HandlePush, github.PushEvent)
+	return webhooks.Run(hook, ":"+strconv.Itoa(GitHookPort), fmt.Sprintf("/%v/handle/push", apiVersion))
+}
+func matchingBranch(inputBranch string, payloadBranch string) bool {
+	if inputBranch == payloadBranch {
+		return true
+	}
+	if inputBranch == "" && payloadBranch == "master" {
+		return true
+	}
+	return false
+}
+
+func (s *gitHookServer) findMatchingPipelineInputs(payload github.PushPayload) (pipelines []*pps.PipelineInfo, inputs []*pps.GithubInput, err error) {
+	var walkInput func(*pps.Input)
+	payloadBranch := getBranch(payload.Ref)
+	pipelines, err = s.client.ListPipeline()
+	if err != nil {
+		return nil, nil, err
+	}
+	for _, pipelineInfo := range pipelines {
+		pps.VisitInput(pipelineInfo.Input, func(input *pps.Input) {
+			if input.Github != nil {
+				if input.Github.URL == payload.Repository.CloneURL && matchingBranch(input.Github.Branch, payloadBranch) {
+					inputs = append(inputs, input.Github)
+				}
+			}
+			var inputs []*pps.Input
+			if input.Cross != nil {
+				inputs = input.Cross
+			}
+			if input.Union != nil {
+				inputs = input.Union
+			}
+			for _, input := range inputs {
+				walkInput(input)
+			}
```

Comment:
I don't understand what this code is doing at all.

Reply:
Ah oops, thats a leftover from where I made my own recursive walker before I found the `VisitInput` function.